tiledb/sm/misc/cancelable_tasks.h

```diff
@@ -0,0 +1,147 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the CancelableTasks class.
+ */
+
+#ifndef TILEDB_CANCELABLE_TASKS_H
+#define TILEDB_CANCELABLE_TASKS_H
+
+#include <functional>
+#include <iostream>
+
+#include "tiledb/sm/misc/status.h"
+#include "tiledb/sm/misc/thread_pool.h"
+
+namespace tiledb {
+namespace sm {
+
+class CancelableTasks {
+ public:
+  /**
+   * Constructor.
+   */
+  CancelableTasks()
+      : outstanding_tasks_(0)
+      , should_cancel_(false) {
+  }
+
+  /**
+   * Destructor.
+   */
+  ~CancelableTasks() = default;
+
+  /**
+   * Enqueue a new task to be executed on the specified thread pool.
+   *
+   * @param function Task to be executed.
+   * @param function Optional routine to execute on cancelation.
+   * @return Future for the return value of the task.
+   */
+  std::future<Status> enqueue(
+      ThreadPool* const thread_pool,
+      std::function<Status()>&& fn,
+      std::function<void()>&& on_cancel = nullptr) {
+    std::function<Status()> wrapped_fn =
+        std::bind(&CancelableTasks::fn_wrapper, this, fn, on_cancel);
+
+    std::future<Status> task = thread_pool->enqueue(std::move(wrapped_fn));
+    if (task.valid()) {
+      std::unique_lock<std::mutex> lck(outstanding_tasks_mutex_);
+      ++outstanding_tasks_;
+    }
+
+    return task;
+  }
+
+  /**
+   * Waits for all enqueued tasks to cancel. If a task is already running, it
+   * will run to completion.
+   */
+  void cancel_all_tasks() {
+    {
+      std::unique_lock<std::mutex> lck(outstanding_tasks_mutex_);
+      should_cancel_ = true;
+    }
+
+    // Wait for all outstanding tasks to cancel.
+    {
+      std::unique_lock<std::mutex> lck(outstanding_tasks_mutex_);
+      outstanding_tasks_cv_.wait(
+          lck, [this]() { return outstanding_tasks_ == 0; });
+      should_cancel_ = false;
+    }
+  }
+
+ private:
+  /**
+   * The wrapped task decorator. If all tasks have been cancelled, they will
+   * short-circuit here with an appropriate return value.
+   *
+   * @param function Task to be executed.
+   * @param function Optional routine to execute on cancelation.
+   * @return Status The returned status from 'fn', or a non-OK status if tasks
+   * were cancelled.
+   */
+  Status fn_wrapper(
+      const std::function<Status()>& fn,
+      const std::function<void()>& on_cancel) {
+    std::unique_lock<std::mutex> lck(outstanding_tasks_mutex_);
```

Comment:
Doesn't this prevent cancellable tasks from executing in parallel, since you're holding this through execution of `fn()`?

Reply:
Oops -- I hope this is the most embarrassing code I write all year. I've released the lock here for both fn() and on_cancel. Note that this complicates the path a little more because we must check to see if 'should_cancel_' changed between the time we released and reacquired the lock.

I re-ran the unit tests successfully.